### PR TITLE
Fix up Dashboard Plugin props

### LINKED ISSUE #392: Error trying to use Dashboard plugins from TypeScript
**Description**
When trying to use Dashboard plugins from a TypeScript file, get an error about props not being filled in. These props are filled in later by the Dashboard itself.

**Steps to reproduce**
1. Create a file App.tsx
2. Add a dashboard with some plugins, e.g.:
```
      <Dashboard>
        <GridPlugin hydrate={hydrateGrid} />
        <LinkerPlugin />
      </Dashboard>
```

**Expected results**
2. Should compile fine

**Actual results**
2. Get a TS error:
```
Type '{}' is missing the following properties from type 'DashboardPluginComponentProps': id, layout, panelManager, registerComponentts(2739)
```

**Additional details and attachments**
image

### PR DESCRIPTION
**Pull request by mofojed**
- Make dashboard props for plugin optional
- Add an assertIsDashboardPluginProps method to assert optional dashboard props
- Add unit tests to verify

Fixes #392

### CODE REVIEW

File: packages/dashboard-core-plugins/src/GridPlugin.tsx [hunk 1]
```diff
@@ -11,22 +12,24 @@ import shortid from 'shortid';
...
-}: GridPluginComponentProps): JSX.Element => {
+export const GridPlugin = (props: GridPluginProps): JSX.Element => {
+  assertIsDashboardPluginProps(props);
+  const {
+    getDownloadWorker,
+    loadPlugin,
+    id,
+    layout,
+    registerComponent,
+    hydrate = DashboardUtils.hydrate,
```
Comment (mattrunyon) [inline packages/dashboard-core-plugins/src/GridPlugin.tsx:30]: hydrate should be optional if it has a default value

File: packages/dashboard-core-plugins/src/PandasPlugin.tsx [hunk 2]
```diff
@@ -11,16 +12,18 @@ import shortid from 'shortid';
...
-export const PandasPlugin = ({
-  hydrate = DashboardUtils.hydrate,
-  id,
-  layout,
-  registerComponent,
-}: PandasPluginComponentProps): JSX.Element => {
+export const PandasPlugin = (props: PandasPluginProps): JSX.Element => {
+  assertIsDashboardPluginProps(props);
+  const {
+    hydrate = DashboardUtils.hydrate,
```
Comment (mattrunyon) [inline packages/dashboard-core-plugins/src/PandasPlugin.tsx:22]: Not an optional prop

File: packages/dashboard-core-plugins/src/ChartPlugin.tsx [hunk 3]
```diff
@@ -11,16 +11,13 @@ import shortid from 'shortid';
...
 
-export const ChartPlugin = ({
-  id,
-  layout,
-  registerComponent,
-  hydrate = DashboardUtils.hydrate,
-}: ChartPluginComponentProps): JSX.Element => {
+export const ChartPlugin = (props: ChartPluginProps): JSX.Element => {
+  assertIsDashboardPluginProps(props);
+  const { id, layout, registerComponent, hydrate } = props;
```
Comment (mattrunyon) [inline packages/dashboard-core-plugins/src/ChartPlugin.tsx:20]: In the other files you have hydrate w/ a default value. Just make it consistent w/ the others

File: packages/dashboard-core-plugins/src/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    getState: jest.fn(() => ({
+      dashboardData: {},
+      workspace: {
+        data: {
+          settings: {},
+        },
+      },
+    })),
+    subscribe: jest.fn(() => undefined),
+  } as unknown) as Store;
```
Comment (mattrunyon) [inline packages/dashboard-core-plugins/src/index.test.tsx:28]: I feel like this should be mocked somewher

File: packages/dashboard/src/DashboardPlugin.ts [hunk 5]
```diff
@@ -84,3 +84,29 @@ export interface DashboardPlugin {
...
+ * @returns True if the props are valid DashboardPluginComponentProps, false otherwise
+ */
+export function isDashboardPluginProps(
+  props: Partial<DashboardPluginComponentProps>
+): props is DashboardPluginComponentProps {
+  return (
+    props.id !== undefined &&
+    props.layout !== undefined &&
+    props.panelManager !== undefined &&
+    props.registerComponent !== undefined
```
Comment (mattrunyon) [inline packages/dashboard/src/DashboardPlugin.ts:100]: Can any of these be more specific? `typeof id !== 'string'`? `typeof registerComponent !== 'function'`?

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/dashboard-core-plugins/src/GridPlugin.tsx (hunk 1 above)
Comment (mofojed) [reply packages/dashboard-core-plugins/src/GridPlugin.tsx:30]: This should be a required param, I'll take out the default value assignment

File: packages/dashboard/src/DashboardPlugin.ts (hunk 5 above)
Comment (mofojed) [reply packages/dashboard/src/DashboardPlugin.ts:100]: Yes

File: packages/dashboard-core-plugins/src/index.test.tsx (hunk 4 above)
Comment (mofojed) [reply packages/dashboard-core-plugins/src/index.test.tsx:28]: It is now mocked

**Review by mattrunyon — APPROVED** [review]

**Pull request merged**